libs/libcds/include/hpx/libcds/hpx_tls_manager.hpp

```diff
@@ -0,0 +1,252 @@
+//  Copyright (c) 2020 Weile Wei
+//
+//  SPDX-License-Identifier: BSL-1.0
+//  Distributed under the Boost Software License, Version 1.0.
+//  (See accompanying file LICENSE_1_0.txt or copy at
+//  http://www.boost.org/LICENSE_1_0.txt)
+
+#pragma once
+
+#include <hpx/config.hpp>
+#include <hpx/modules/errors.hpp>
+#include <hpx/modules/functional.hpp>
+//
+#include <cds/gc/details/hp_common.h>
+#include <cds/gc/hp.h>
+#include <cds/init.h>
+#include <cds/threading/details/cxx11_manager.h>
+#include <cds/urcu/general_buffered.h>
+
+#include <atomic>
+#include <cstddef>
+#include <iostream>
+#include <string>
+
+/// \cond NODETAIL
+namespace cds { namespace gc { namespace hp { namespace details {
+
+    class HPXDataHolder
+    {
+    public:
+        CDS_EXPORT_API static CDS_EXPORT_API thread_data* getTLS();
+        CDS_EXPORT_API static CDS_EXPORT_API void setTLS(thread_data*);
+        CDS_EXPORT_API static CDS_EXPORT_API generic_smr<HPXDataHolder>*
+        getInstance();
+        CDS_EXPORT_API static CDS_EXPORT_API void setInstance(
+            generic_smr<HPXDataHolder>* new_instance);
+    };
+
+}}}}    // namespace cds::gc::hp::details
+
+namespace hpx { namespace cds {
+
+    namespace detail {
+        /// \cond NOINTERNAL
+        using get_num_concurrent_hazard_pointer_threads_type =
+            hpx::util::function_nonser<std::size_t()>;
+        HPX_EXPORT void set_get_num_concurrent_hazard_pointer_threads(
+            get_num_concurrent_hazard_pointer_threads_type f);
+        HPX_EXPORT std::size_t get_num_concurrent_hazard_pointer_threads();
+        /// \endcond
+    }    // namespace detail
+
+    enum class smr_t
+    {
+        hazard_pointer_hpxthread,
+        hazard_pointer_stdthread,
+        rcu
+    };
+
+    // this wrapper will initialize libCDS
+    struct libcds_wrapper
+    {
+        friend struct hpxthread_manager_wrapper;
+        friend struct stdthread_manager_wrapper;
+
+        libcds_wrapper(smr_t smr_type = smr_t::hazard_pointer_hpxthread,
+            std::size_t hazard_pointer_count = 1,
+            std::size_t max_thread_count =
+                detail::get_num_concurrent_hazard_pointer_threads(),
+            std::size_t max_retired_pointer_count = 16)
```

Comment:
I've added the callback here. My question now is if those other parameters should have configuration options as well or not? Is this the appropriate order for the default parameters (i.e. which ones are the user most likely to change)?

Reply:
FWIW, having three `size_t` arguments is highly confusing. Nobody will be able to remember what sequence those have to be supplied in.

Reply:
@msimberg All these parameters should be changed according to the cds container you are using, and the number of threads you want to use. 

@hkaiser if not `size_t` what type should we use then?

Reply:
@weilewei this is a tough question. `size_t` is the right type to use. The problem is that it's difficult to pass the values in a correct sequence. One possible way to enforce the user has to use names would be to define three structs, like:
```
struct hazard_pointer_count 
{ 
    explicit hazard_pointer_count(std::size_t value) : value_(value) {}

private:
    std::size_t value_;
};
```
and use those as:
```
struct libcds_wrapper
{
    libcds_wrapper(smr_t smr_type = smr_t::hazard_pointer_hpxthread,
        hazard_pointer_count count = hazard_pointer_count(1),
        max_thread_count max_threads = max_thread_count(
            detail::get_num_concurrent_hazard_pointer_threads()),
        max_retired_pointer_count retired_count = max_retired_pointer_count(16))
    {
        // ...
    }
};
```